SpatialGDK/Source/SpatialGDK/Private/Interop/Connection/SpatialTraceEventDataBuilder.cpp

```diff
@@ -0,0 +1,277 @@
+const char* FSpatialTraceEventName::ReceiveCommandResponseEventName = GDK_EVENT_NAMESPACE "receive_command_response";
+const char* FSpatialTraceEventName::ReceiveCommandRequestEventName = GDK_EVENT_NAMESPACE "receive_command_request";
+const char* FSpatialTraceEventName::ApplyRPCEventName = GDK_EVENT_NAMESPACE "apply_rpc";
+const char* FSpatialTraceEventName::SendRPCEventName = GDK_EVENT_NAMESPACE "send_rpc";
+const char* FSpatialTraceEventName::QueueRPCEventName = GDK_EVENT_NAMESPACE "queue_rpc";
+const char* FSpatialTraceEventName::ReceieveRPCEventName = GDK_EVENT_NAMESPACE "receive_rpc";
+const char* FSpatialTraceEventName::MergeSendRPCEventName = GDK_EVENT_NAMESPACE "merge_send_rpcs";
+const char* FSpatialTraceEventName::ApplyCrossServerRPCEventName = GDK_EVENT_NAMESPACE "apply_cross_server_rpc";
+const char* FSpatialTraceEventName::SendCommandResponseEventName = GDK_EVENT_NAMESPACE "send_command_response";
+const char* FSpatialTraceEventName::SendCommandRequestEventName = GDK_EVENT_NAMESPACE "send_command_request";
+const char* FSpatialTraceEventName::SendCrossServerRPCEventName = GDK_EVENT_NAMESPACE "send_cross_server_rpc";
+const char* FSpatialTraceEventName::ReceiveCrossServerRPCEventName = GDK_EVENT_NAMESPACE "receive_cross_server_rpc";
+const char* FSpatialTraceEventName::MergePropertyUpdateEventName = GDK_EVENT_NAMESPACE "merge_property_update";
+const char* FSpatialTraceEventName::ReceiveCreateEntityEventName = GDK_EVENT_NAMESPACE "receive_create_entity";
+const char* FSpatialTraceEventName::ReceiveRemoveEntityEventName = GDK_EVENT_NAMESPACE "receive_remove_entity";
+const char* FSpatialTraceEventName::AuthorityChangeEventName = GDK_EVENT_NAMESPACE "authority_change";
+
+// ---- FStringCache ----
+
+int32 FSpatialTraceEventDataBuilder::FStringCache::CombineStrings(const char* A, const char* B)
+{
+	if (NextIndex >= BufferSize)
+	{
+		return BufferSize - 1;
+	}
+
+	int32 InsertIndex = NextIndex;
+	int32 InitialRemainingSize = BufferSize - NextIndex;
+	FCStringAnsi::Strncpy(&Buffer[NextIndex], A, InitialRemainingSize);
+
+	int32 CharSize = sizeof(char);
+	int32 ALength = FCStringAnsi::Strlen(A);
+	int32 RemainingSize = InitialRemainingSize - ALength;
+	RemainingSize = FMath::Max(0, RemainingSize);
+
+	if (RemainingSize > CharSize)
+	{
+		FCStringAnsi::Strncpy(&Buffer[NextIndex + ALength], B, RemainingSize);
+	}
+
+	int32 BLength = FCStringAnsi::Strlen(B);
+	RemainingSize = RemainingSize - BLength - CharSize;
+	RemainingSize = FMath::Max(0, RemainingSize);
+
+	NextIndex += InitialRemainingSize - RemainingSize;
+	return InsertIndex;
+}
+
+int32 FSpatialTraceEventDataBuilder::FStringCache::AddString(const char* String)
+{
+	if (NextIndex >= BufferSize)
+	{
+		return BufferSize - 1;
+	}
+
+	int32 InsertIndex = NextIndex;
+	int32 InitialRemainingSize = BufferSize - NextIndex;
+	FCStringAnsi::Strncpy(&Buffer[NextIndex], String, InitialRemainingSize);
+
+	int32 CharSize = sizeof(char);
+	int32 StringLength = FCStringAnsi::Strlen(String);
```

Comment:
Does this make more sense if we add 1 to strlen rather than subtract CharSize below? 

Also the FMath::Max is never required because RemainingSize is always >= 0 (the largest NextIndex is BufferSize-1 which leaves RemainingSize at 0)

Reply:
Yeah if you prefer.

RemainingSize can be less than 0. InitialRemainingSize maxes out at BufferSize (1000). StringLength can be any posative number.